[extension/ecsobserver] Init service client and mock

pingleig opened the pull request:
**Description:**

- Implement get all tasks
- Package `ecsmock` mocks ECS service API for a single cluster. Might move it into `internal/aws` once #3188 is done.

**Link to tracking Issue:**

- Feature request #1395 

Pending PRs

- Docker Label Matcher https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3276
- Task IP and Port https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3133

Previous PRs

- Target and Task definition (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2939
- Config PR (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377
- Original PR (closed) #2734 

**Testing:** 

unit test

**Documentation:** 

No new doc. See [existing doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md)

pingleig committed:
ext: ecsobserver Init service client and mock

- Implement get all tasks
- Pacakge `ecsmock` provides ECS service API for a single cluster.

pingleig commented:
cc @Aneurysm9 @jrcamp

mxiamxia commented on extension/observer/ecsobserver/ecsmock/service.go:
```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+// API Start
```
Pls help to enhance this comment.

pingleig replied:
This is a marker for making start and end of some block so it's easier for people to navigate the code. And if you have extra config in editor, it can helps folding the code.

mxiamxia commented on extension/observer/ecsobserver/ecsmock/service.go:
```diff
@@ -0,0 +1,195 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecsmock
```
Pls provide go doc comments for this package

pingleig replied:
The package level doc is in `pkg.go` I think I can rename the file to `doc.go`

mxiamxia commented on extension/observer/ecsobserver/fetcher.go:
```diff
@@ -0,0 +1,88 @@
+	Cluster string
+	Region  string
+
+	// test overrides
+	ecs ecsClient
```
change to allow
```suggestion
	// custom ecsClient Fetcher overrides
	ecs ecsClient
```

pingleig replied:
There will be two overrides, I deleted the ec2 one to make the PR smaller. 

I think change the  var name to `ecsTestOverride` or `ecsTestClient` is easier compared w/ adding comment for each client.

anuraaga commented on extension/observer/ecsobserver/ecsmock/service.go:
```diff
@@ -0,0 +1,195 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecsmock
```
Can you move this into an `ecsobserver/internal/ecsmock` package?

pingleig replied:
done

anuraaga commented on extension/observer/ecsobserver/fetcher.go:
```diff
@@ -0,0 +1,88 @@
+	DescribeTasksWithContext(ctx context.Context, input *ecs.DescribeTasksInput, opts ...request.Option) (*ecs.DescribeTasksOutput, error)
+}
+
+type TaskFetcher struct {
```
Can these be private?

anuraaga reviewed (commented):
As we're moving towards GA we need to be careful on keeping exposed API to a minimum. Can you skim through all the current PRs and make sure structs, methods, and constants are all either

- public to implement a collector interface
- private
- Subpackage of `internal`

?

pingleig committed:
ext: ecsobserver Move ecsmock to internal

It should get moved into internal/aws in the future. Keep it under
extension package for now

pingleig commented:
@anuraaga PTAL
